etcdserver/cluster_util.go

```diff
@@ -177,6 +177,120 @@ func getVersions(lg *zap.Logger, cl *membership.RaftCluster, local types.ID, rt
 	return vers
 }
 
+// decideAllowedVersionRange decides the available version range of the cluster that local server can join in;
+// if the downgrade enable status of the cluster is not decided(some server allows downgrade some not),
+// the version window is [localVersion, localVersion];
+// if the downgrade enable status is true, the version window is [localVersion, oneMinorHigher]
+// if the downgrade is not enabled, the version window is [MinClusterVersion, localVersion]
+func decideAllowedVersionRange(enables []bool) (minV *semver.Version, maxV *semver.Version) {
+	minV = semver.Must(semver.NewVersion(version.MinClusterVersion))
+	maxV = semver.Must(semver.NewVersion(version.Version))
+	maxV = &semver.Version{Major: maxV.Major, Minor: maxV.Minor}
+
+	if len(enables) == 0 {
+		return minV, maxV
+	}
+	enable := enables[0]
+	for _, e := range enables {
+		// if the downgrade enable status of the cluster is not decided,
```

Comment:
Can we explore the idea of having "/downgrade/enable endpoint serving linearizable data? This way it is much easier to decide allowed server version range. And the logic and reasoning will be cleaner. For a given cluster, the downgrade enable status is never "undecided", we cannot decide the status here because we are reading possibly stale data from different members.

Reply:
Good point! Updated the downgrade handler.